Build tests with -Wunused-function with cmake

gilles-peskine-arm opened the pull request:
This was previously done for make in #2053. To avoid builds passing with one build system and failing with the other, align cmake with make.

Backport: I think not. It could break the way users run unit tests in a custom configuration. Having fewer warnings in the LTS version is not going to make future backports harder, so there's no significant advantage to backporting.

AndrzejKurek reviewed (approved):
I wouldn't say that this PR completely aligns cmake and make, but yes, the result is more desirable:
default cmake build flags, directories/includes/object files omitted:
`-D_POSIX_C_SOURCE=200809L -Wall -Wextra -Wwrite-strings -Wformat=2 -Wno-format-nonliteral -Wvla -Wlogical-op -Wshadow -Wformat-signedness -Wformat-overflow=2 -Wformat-truncation -Werror`
default make build flags, directories/includes/object files omitted:
`-Wall -Wextra -Wformat=2 -Wno-format-nonliteral -D_FILE_OFFSET_BITS=64 -D_POSIX_C_SOURCE=200809L -O2 test_suite_pkcs5.c `

gilles-peskine-arm commented:
> I wouldn't say that this PR completely aligns cmake and make

Indeed it doesn't, but that wasn't my objective. I encountered the unused function warning in `test_suite_ssl` when building with make and I realized the CI hadn't complained because it was only building this configuration with cmake, so I went and fixed this particular discrepancy.

There is some legitimate deviation between cmake and make because cmake is aware of the compiler implementation and version. But this one was unintended, we just changed one and forgot the other.
